odpf/shield/v1/shield.proto

```diff
@@ -0,0 +1,496 @@
+    option (google.api.http) = {
+      put: "/v1/projects/{id}",
+      body: "body"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      tags: "Project";
+      summary: "Update Project by ID";
+    };
+  }
+
+}
+
+message CreateUserRequestBody {
+  string name = 1 [(validate.rules).string.pattern = "^[A-Za-z0-9_-]+$"];
+  string email = 2 [(validate.rules).string.email = true];
+  google.protobuf.Struct metadata = 3;
+}
+
+message CreateUserRequest {
+  CreateUserRequestBody body = 1;
+}
+
+message User {
+  string id = 1;
+  string name = 2 [(validate.rules).string.pattern = "^[A-Za-z0-9_-]+$"]; 
+  string slug = 3;
+  string email = 4 [(validate.rules).string.email = true]; 
+  google.protobuf.Struct metadata = 5;
+  google.protobuf.Timestamp createdAt = 6;
+  google.protobuf.Timestamp updatedAt = 7;
+}
+
+message UserResponse {
+  User user = 1;
+}
+
+message UpdateUserRequest {
+  string id = 1;
+  CreateUserRequestBody body = 2;
+}
+
+message GetUserRequest {
+  string id = 1;
+}
+
+message GetCurrentUserRequest {
+}
+
+message GetAllUsersRequest {
+  google.protobuf.Struct fields = 1;
+}
+
+message GetAllUsersResponse {
+  repeated User users = 1;
+}
+
+
+message CreateGroupRequestBody {
+  string name = 1 [(validate.rules).string.pattern = "^[A-Za-z0-9_-]+$"];
+  string slug = 2;
+  google.protobuf.Struct metadata = 3;
+}
+
+message CreateGroupRequest {
+  CreateGroupRequestBody body = 1;
+}
+
+message Admin {
+  string id = 1;
+  string name = 2;
+}
+
+message Group {
+  string id = 1;
+  string name = 2 [(validate.rules).string.pattern = "^[A-Za-z0-9_-]+$"];
+  string slug = 3;
+  repeated Admin admins = 4;
+  bool isMember = 5;
+  int32 memberCount = 6;
+  string orgId = 7;
+  google.protobuf.Struct metadata = 8;
+  google.protobuf.Timestamp createdAt = 9;
+  google.protobuf.Timestamp updatedAt = 10;
+}
+
+message GroupResponse {
+  Group group = 1;
+}
+
+message GetGroupUsersResponse {
+  repeated User users = 1;
+}
+
+message UpdateGroupRequest {
+  string id = 1;
+  CreateGroupRequestBody body = 2;
+}
+
+message UpdateCurrentUserRequest {
+  CreateGroupRequestBody body = 1;
+}
+
+message GetGroupRequest {
+  string id = 1;
+}
+
+message GetAllGroupsRequest {
+  string userId = 1;
+//  string orgId = 2;
+//  string org = 3;
+}
+
+message GetAllGroupsResponse {
+  repeated Group groups = 1;
+}
+
+message Role {
+  string id = 1;
+  string name = 2 [(validate.rules).string.pattern = "^[A-Za-z0-9_-]+$"];
+  repeated string tags = 3;
+  repeated string actions = 4;
+  google.protobuf.Struct metadata = 5;
+  google.protobuf.Timestamp createdAt = 6;
+  google.protobuf.Timestamp updatedAt = 7;
+}
+
+message CreateRoleBody {
+  string name = 1 [(validate.rules).string.pattern = "^[A-Za-z0-9_-]+$"];
+  repeated string tags = 2;
+  repeated string actions = 3;
+  google.protobuf.Struct metadata = 4;
+}
+
+message CreateRoleRequest {
+  CreateRoleBody body = 1;
```

Comment:
If this `CreateRoleBody` is tightly part of `CreateRoleRequest` why not we make it a sub-message of this request? If we don't think we want to tightly couple them then just embed these fields directly to `CreateRoleRequest`?

Reply:
`CreateRoleBody` will also be used by `UpdateRoleRequest`.

Reply:
Can't we utilize the `message Role` itself instead of `CreateRoleBody` ?